src/data-table/utils/selection.tsx

```diff
@@ -1,39 +1,46 @@
-import { MouseEvent } from "react";
-import _ from "lodash";
 import i18n from "@dhis2/d2-i18n";
+import _ from "lodash";
+import { MouseEvent } from "react";
+import {
+    ReferenceObject,
+    TableAction,
+    TableNotification,
+    TablePagination,
+    TableSelection,
+} from "../types";
 
-import { ReferenceObject, TableAction, TablePagination, TableNotification } from "../types";
-
-export function updateSelection<T extends ReferenceObject>(selected: string[], row: T) {
-    let newSelected: string[] = [];
-    const selectedIndex = selected.indexOf(row.id);
+export function updateSelection<T extends ReferenceObject>(selected: TableSelection[], row: T) {
+    const selectedIndex = _.findIndex(selected, { id: row.id });
 
     if (selectedIndex === -1) {
-        newSelected = newSelected.concat(selected, row.id);
+        return [...selected, { id: row.id }];
     } else if (selectedIndex === 0) {
-        newSelected = newSelected.concat(selected.slice(1));
+        return selected.slice(1);
     } else if (selectedIndex === selected.length - 1) {
-        newSelected = newSelected.concat(selected.slice(0, -1));
+        return selected.slice(0, -1);
     } else if (selectedIndex > 0) {
-        newSelected = newSelected.concat(
-            selected.slice(0, selectedIndex),
-            selected.slice(selectedIndex + 1)
-        );
+        return [...selected.slice(0, selectedIndex), ...selected.slice(selectedIndex + 1)];
+    } else {
+        return [];
     }
-
-    return newSelected;
 }
 
 export function isEventCtrlClick(event: MouseEvent<unknown>) {
     return event && event.ctrlKey;
 }
 
-export function getActionRows<T extends ReferenceObject>(row: T, rows: T[], selection: string[]) {
-    const rowInSelection = !!selection.find(id => row.id === id);
+export function getActionRows<T extends ReferenceObject>(
+    selectedRow: T,
+    allRows: T[],
+    selection: TableSelection[]
+) {
+    const isRowInSelection = !!_.find(selection, { id: selectedRow.id });
+    const selectedRows = _(selection)
+        .map(({ id }) => _.find(allRows, ["id", id]))
```

Comment:
a find nested in a map is O(n*m), which can get quite large. Typical solution: process inner data structure so this predicates are O(1)

Reply:
Simplified with:

```
    const selectedRows = _(allRows)
        .intersectionBy(selection, "id")
        .value();
```

Thanks @tokland!